src/vm/synch.cpp

```diff
@@ -590,6 +590,376 @@ DWORD CLRSemaphore::Wait(DWORD dwMilliseconds, BOOL alertable)
+#endif // FEATURE_PAL
+    if (h == nullptr)
+    {
+        ThrowOutOfMemory();
+    }
+
+    m_handle = h;
+    m_counts.signalCount = initialSignalCount;
+    INDEBUG(m_maximumSignalCount = maximumSignalCount);
+}
+
+void CLRLifoSemaphore::Close()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    if (m_handle == nullptr)
+    {
+        return;
+    }
+
+    CloseHandle(m_handle);
+    m_handle = nullptr;
+}
+
+bool CLRLifoSemaphore::WaitForSignal(DWORD timeoutMs)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        SO_TOLERANT;
+    }
+    CONTRACTL_END;
+
+    _ASSERTE(timeoutMs != 0);
+    _ASSERTE(m_handle != nullptr);
+    _ASSERTE(m_counts.waiterCount != (UINT16)0);
+
+    while (true)
+    {
+        // Wait for a signal
+        BOOL waitSuccessful;
+        {
+#ifdef FEATURE_PAL
+            // Do a prioritized wait to get LIFO waiter release order
+            DWORD waitResult = PAL_WaitForSingleObjectPrioritized(m_handle, timeoutMs);
+            _ASSERTE(waitResult == WAIT_OBJECT_0 || waitResult == WAIT_TIMEOUT);
+            waitSuccessful = waitResult == WAIT_OBJECT_0;
+#else // !FEATURE_PAL
+            // I/O completion ports release waiters in LIFO order, see
+            // https://msdn.microsoft.com/en-us/library/windows/desktop/aa365198(v=vs.85).aspx
+            DWORD numberOfBytes;
+            ULONG_PTR completionKey;
+            LPOVERLAPPED overlapped;
+            waitSuccessful = GetQueuedCompletionStatus(m_handle, &numberOfBytes, &completionKey, &overlapped, timeoutMs);
+            _ASSERTE(waitSuccessful || GetLastError() == WAIT_TIMEOUT);
+            _ASSERTE(overlapped == nullptr);
+#endif // FEATURE_PAL
+        }
+
+        // Unregister the waiter if this thread will not be waiting anymore, and try to acquire the semaphore
+        Counts counts = m_counts.VolatileLoad();
+        while (true)
+        {
+            _ASSERTE(counts.waiterCount != (UINT16)0);
+            Counts newCounts = counts;
+            if (counts.signalCount != 0)
+            {
+                --newCounts.signalCount;
+                --newCounts.waiterCount;
+            }
+            else if (!waitSuccessful)
+            {
+                --newCounts.waiterCount;
+            }
+
+            // This waiter has woken up and this needs to be reflected in the count of waiters signaled to wake. Since we don't
+            // have thread-specific signal state, there is not enough information to tell whether this thread woke up because it
+            // was signaled. For instance, this thread may have timed out and then we don't know whether this thread also got
+            // signaled. So in any woken case, decrement the count if possible. As such, timeouts could cause more waiters to
+            // wake than necessary.
+            if (counts.countOfWaitersSignaledToWake != (UINT8)0)
+            {
+                --newCounts.countOfWaitersSignaledToWake;
+            }
+
```

Comment:
Is it possible / likely that newCounts == counts?  I'm wondering if there's any value in checking whether they're the same before doing the interlocked operation.

Reply:
In this case checking before worries me a bit regarding countOfWaitersSignaledToWake. Checking before and exiting before the interlocked operation would be speculative and if it happens to read an old value and sees a zero, missing a write in Release, the count could be greater than it should be and that could result in a waiter not being woken.

Reply:
Thought about it some more, I don't think it would be wrong for the reason I gave above. If new and old counts are the same and it incorrectly sees a zero for countOfWaitersSignaledToWake, it must be because another thread decremented that count to zero and yet another thread incremented it back to nonzero. The thread that incremented the count to nonzero will wake another thread so it's ok for this thread to miss the decrement, the next woken thread will take care of that.

I don't see an issue with checking/breaking early here at the moment, but I prefer to err on the safe side and not do so to avoid any potential problems that I may have missed.

It's not very likely for the counts to be the same because typically every woken thread has to decrement countOfWaitersSignaledToWake. If a waiter timed out in-between and another thread acquired the semaphore it could cause the counts to be the same. Waiters waking up also wouldn't typically be on the hot path, there could be significant delays between bursts of work items to cause threads to wait before waking up for instance, but the delay would mask any gain that this would give.

Reply:
Ok. Thanks.